File: frontend/src/types/immigration.ts
export interface LanguageScores {
  speaking: number;
  listening: number;
  reading: number;
  writing: number;
}

export interface CRSInput {
  age: number;
  has_spouse: boolean;
  education_level: string;
  first_language: LanguageScores;
  second_language?: LanguageScores | null;
  canadian_work_experience_years: number;
  foreign_work_experience_years: number;
  spouse_education_level?: string | null;
  spouse_language?: LanguageScores | null;
  spouse_canadian_work_years: number;
  has_certificate_of_qualification: boolean;
  has_provincial_nomination: boolean;
  job_offer_noc_level?: string | null;
  canadian_education_years: number;
  has_sibling_in_canada: boolean;
  french_clb7_plus: boolean;
  english_clb5_plus: boolean;
}

export interface CRSBreakdown {
  core_human_capital: number;
  spouse_factors: number;
  skill_transferability: number;
  additional_points: number;
}

export interface CRSResult {
  total_score: number;
  breakdown: CRSBreakdown;
  max_score: number;
}

export interface ImmigrationProgram {
  id: string;
  name: string;
  province: string;
  program_type: string;
  min_crs_score: number | null;
  requirements: Record<string, unknown>;
  active: boolean;
}

export interface PathwayMatch {
  program: ImmigrationProgram;
  eligible: boolean;
  match_score: number;
  notes: string[];
}

export interface ChecklistItem {
  id: string;
  title: string;
  description: string;
  category: string;
  order: number;
  is_completed: boolean;
}

export interface EEDraw {
  id: string;
  draw_number: number;
  draw_date: string;
  min_score: number;
  invitations: number;
  program_type: string;
}

File: backend/app/schemas/immigration.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Language helpers
# ---------------------------------------------------------------------------

class LanguageScores(BaseModel):
    speaking: int = Field(..., ge=0, le=12)
    listening: int = Field(..., ge=0, le=12)
    reading: int = Field(..., ge=0, le=12)
    writing: int = Field(..., ge=0, le=12)


# ---------------------------------------------------------------------------
# CRS calculator schemas
# ---------------------------------------------------------------------------

class CRSInput(BaseModel):
    age: int
    has_spouse: bool = False
    education_level: str
    first_language: LanguageScores
    second_language: LanguageScores | None = None
    canadian_work_experience_years: int = 0
    foreign_work_experience_years: int = 0
    spouse_education_level: str | None = None
    spouse_language: LanguageScores | None = None
    spouse_canadian_work_years: int = 0
    has_certificate_of_qualification: bool = False
    has_provincial_nomination: bool = False
    job_offer_noc_level: str | None = Field(
        default=None,
        description="NOC level of arranged employment: '00', '0ab', 'other', or None",
    )
    canadian_education_years: int = 0
    has_sibling_in_canada: bool = False
    french_clb7_plus: bool = False
    english_clb5_plus: bool = True


class CRSBreakdown(BaseModel):
    core_human_capital: int
    spouse_factors: int
    skill_transferability: int
    additional_points: int


class CRSResult(BaseModel):
    total_score: int
    breakdown: CRSBreakdown
    max_score: int = 1200


# ---------------------------------------------------------------------------
# Immigration program / pathway schemas
# ---------------------------------------------------------------------------

class ImmigrationProgramResponse(BaseModel):
    id: UUID
    name: str
    province: str
    program_type: str
    min_crs_score: int | None
    requirements: dict
    active: bool

    class Config:
        from_attributes = True


class PathwayMatchResponse(BaseModel):
    program: ImmigrationProgramResponse
    eligible: bool
    match_score: float
    notes: list[str]


# ---------------------------------------------------------------------------
# Post-landing checklist
# ---------------------------------------------------------------------------

class ChecklistItem(BaseModel):
    id: str
    title: str
    description: str
    category: str
    order: int
    is_completed: bool = False


# ---------------------------------------------------------------------------
# Express Entry draw
# ---------------------------------------------------------------------------

class EEDrawResponse(BaseModel):
    id: UUID
    draw_number: int
    draw_date: date
    min_score: int
    invitations: int
    program_type: str

    class Config:
        from_attributes = True
